clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -2075,9 +2076,41 @@ protected DelegateConsumerRecordResult delegateConsumerRecord(
                * For #3 stat counter update will not happen for SOS/EOS message. This should not be a big issue. If needed we can copy some of the stats maintenance
                *   work here.
                *
+               * To address #2 we will periodically produce {@link com.linkedin.venice.kafka.protocol.SyncOffset}
+               * control message to VT and upon consumption of those messages in the drainer we will update the offset.
+               *
                * So in summary NO further processing is needed SOS/EOS received from RT topics. Just silently drop the message here.
                * We should not return false here.
                */
+              if (consumerRecord.getTopicPartition().getPubSubTopic().isRealTime()) {
```

Comment:
This approach is still subject to the requirements that there is data in RT, or we happen to produce SyncOffset at the end of RT.

I wonder if we can find a way to produce SyncOffset regularly without depending on RT, and without introducing race condition (or the readers of this CM is robust to such race).

Could you please check if producer is thread safe? My hunch is yes. If so, can we produce this CM regularly in this function: `checkIngestionProgress`? One potential issue is that the upstream offset produced in this function could be stale; one potential case:
1. share consumer thread produce a data message with upstream 100;
2. ingestion thread get the latest consumed RT offset in function `checkIngestionProgress` and think the current value is 100;
3. before the ingestion thread is able to send out SyncOffset, share consumer thread sends out a new data message with upstream offset 101
4. ingestion thread send out SyncOffset with upstream offset 100
When it happens, the reader of SyncOffset just need to ignore offset 100 since it's smaller than the checkpoint upstream offset 101 which is correct.

Please help double check if it makes sense. The drawback of keeping the regularly producing logic here is that - let's say the end offset of RT is 102, the last 2 messages are EOS, and consumer threads already produce a SyncOffset CM at upstream offset 100, then we will end up in the same situation, since it wouldn't take longer than 15 minutes to process the rest 2 messages.

@sushantmane Could you please also double check here? Thanks!

Reply:
I think this is doable, will try to implement it this way and control the behavior via a server config for smooth rollout.